### PR TITLE
Wip rgw kafka timeout

### PR DESCRIPTION
**Pull request by TRYTOBE8TME**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_kafka.cc [hunk 1]
```diff
@@ -74,6 +73,7 @@ struct connection_t {
   const boost::optional<std::string> ca_location;
   const std::string user;
   const std::string password;
+  utime_t timestamp;
```
Comment (yuvalif) [inline src/rgw/rgw_kafka.cc:76]: please initialize to zero

File: src/rgw/rgw_kafka.cc [hunk 2]
```diff
@@ -416,13 +418,10 @@ class Manager {
...
-        // delete the connection if marked for deletion
-        if (conn->marked_for_deletion) {
-          ldout(conn->cct, 10) << "Kafka run: connection is deleted" << dendl;
-          conn->destroy(STATUS_CONNECTION_CLOSED);
-          std::lock_guard lock(connections_lock);
-          // erase is safe - does not invalidate any other iterator
-          // lock so no insertion happens at the same time
+
+        // Checking the connection idlesness
+        if(conn->timestamp.sec() + 30 < ceph_clock_now()) {
```
Comment (yuvalif) [inline src/rgw/rgw_kafka.cc:423]: please add a data member to the Manager called "max_idle_time" and set it to 30 in the ctor (similarly to the "max_connection", see lines 295-297)

**Comment by yuvalif** [comment]
please add a test to the bucket notification tests:
* configure a topic and notification
* upload an object and verify the notification is received
* wait 30 sec
* call netstat from the python code to verify that the TCP connection is down
* upload an object and verify the notification is received

**Comment by yuvalif** [comment]
please update the commit messages to include the tracker fixed by this PR

File: src/test/rgw/bucket_notification/test_bn.py [hunk 3]
```diff
@@ -2224,6 +2224,137 @@ def test_ps_s3_persistent_notification_pushback():
...
+    # name is constant for manual testing
+    topic_name = bucket_name+'_topic'
+    # create consumer on the topic
+   
+    task, receiver = create_kafka_receiver_thread(topic_name+'_1')
+    task.start()
+
+    # create s3 topic
+    endpoint_address = 'kafka://' + kafka_server
+    # without acks from broker
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2246]:
```suggestion
    # with acks from broker
```

File: src/test/rgw/bucket_notification/test_bn.py [hunk 4]
```diff
@@ -2224,6 +2224,137 @@ def test_ps_s3_persistent_notification_pushback():
...
+    endpoint_args = 'push-endpoint='+endpoint_address+'&kafka-ack-level=broker'
+    topic_conf1 = PSTopicS3(conn, topic_name+'_1', zonegroup, endpoint_args=endpoint_args)
+    topic_arn1 = topic_conf1.set_config()
+    endpoint_args = 'push-endpoint='+endpoint_address+'&kafka-ack-level=none'
+    topic_conf2 = PSTopicS3(conn, topic_name+'_2', zonegroup, endpoint_args=endpoint_args)
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2251]: why do you need 2 topics in the test?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 5]
```diff
@@ -2224,6 +2224,137 @@ def test_ps_s3_persistent_notification_pushback():
...
+    print('average time for deletion + kafka notification is: ' + str(time_diff*1000/number_of_objects) + ' milliseconds')
+
+    print('wait for 5sec for the messages...')
+    time.sleep(5)
+    receiver.verify_s3_events(keys, exact_match=True, deletions=True, etags=etags)
+
+    print('waiting for 30sec for checking idleness')
+    time.sleep(30)
+
+    os.system("netstat -nnp | grep 9092");
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2308]: can you parse the system call output to verify that the connection is down?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 6]
```diff
@@ -2224,6 +2224,137 @@ def test_ps_s3_persistent_notification_pushback():
...
+
+    time_diff = time.time() - start_time
+    print('average time for deletion + kafka notification is: ' + str(time_diff*1000/number_of_objects) + ' milliseconds')
+
+    print('wait for 5sec for the messages...')
+    time.sleep(5)
+    receiver.verify_s3_events(keys, exact_match=True, deletions=True, etags=etags)
+
+    print('waiting for 30sec for checking idleness')
+    time.sleep(30)
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2306]: probbaly need to sleep little more then 30sec, to make sure we will close the connection

**Review by yuvalif — APPROVED** [review]

**Comment by TRYTOBE8TME** [comment]
Teuthology link: https://pulpito.ceph.com/kapandya-2021-09-01_10:52:22-rgw:notifications-wip-rgw-kafka-timeout-distro-basic-smithi/

**Pull request merged**